### ISSUE TITLE
Are field initializer expressions in const constructor initializer lists considered to be in a const context?

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/sdk#32846]
The analyzer and the CFE disagree about whether the initializer for `x` is in a `const` context.  Which is correct?   My read of https://github.com/dart-lang/sdk/blob/master/docs/language/informal/implicit-creation.md leaves it slightly ambiguous, since `const` variable isn't defined, but assuming that `const` variable means one preceded with `const` then this is not a `const` context.  But it's not clear to me that it shouldn't be?  

```
class A {
  const A();
}

class C {
  final A x;
  const C() : x = A();
}

void main() {

  const C();
}
```

**Comment by eernstg** [comment]
For the current status ([implicit-creation.md](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/implicit-creation.md) and [Lasse's CL on dartLangSpec.tex](https://dart-review.googlesource.com/c/sdk/+/50221)), I'd say that the initializer is not a constant context because we have an explicit list and it's not there (a constant variable has a declaration of the form `const v = e;` and the combination of `final v;` and a constructor initializer to me seems to be clearly different from that).

So maybe we should add it? I think we'd need some more examples. Clearly, `A()` in your example couldn't mean `new A()` because that's not a constant expression even when constructor parameters are considered to be constant expressions, so the example seems to support the assumption that it would be "safe" to let the initializer be a constant context. But we could have the following:

```dart
class B {
  final y;
  const B(this.y);
}

class C {
  final x;
  const C(v) : x = const B(v);
}
```

With the current wording in the spec `const B(v)` would actually be a potentially constant expression, and that would be OK, but it is still rejected by current tools, and it violates the requirement that we have otherwise had (the syntax `const B(v)` is supposed to be in a 1-to-1 correspondence with a run-time entity, and that wouldn't be true if the constructor is called several times with different values for `v`).

Basically, it looks to me like it would be safe to say that the initializing expressions in a constant constructor initializer list are in a constant context, and it might also be quite convenient.

Do you spot any counter examples?

**Comment by lrhn** [comment]
A const constructor initializer expression is currently not defined to be a const context. We do not insert `const` in front of list literals in there. The const contexts are introduced, as listed in the specification for it, around:

* if *e* is an element of a constant list literal, or a key or value of an entry of a constant map literal.
* if *e* is an actual argument of a constant object expression or of a metadata annotation.
* if *e* is the initializing expression of a constant variable declaration.
* if *e* is a switch case expression.
* if *e* is an immediate subexpression of an expression *e*<sub>1</sub> which occurs in a constant context, unless *e*<sub>1</sub> is a throw expression or a function literal.

That list is intended to be exhaustive.  It does not contain all positions where const values are required, and it does not intend to. For example, the list does not include:
* Optional parameter default values.
* Initializers of fields in classes with const constructors.
* Initializer expressions in const constructor initializer lists.

That is deliberate. Each omission has a different reason (but generally boil down to "room for future changes"). 
For default values, it's because we want to keep the door open to making them non-constant in the future, and do not want users to write code that depends on const insertion.
For const class field initializers it's because it's fragile because the const is inferred from something at a distance, there is no syntactically enclosing hint.
For initializer list expressions, it's because we may want to have call-dependent behavior in the future, so a `const` invocation would create a const object, and a `new` invocation would create a new object (and then not require the arguments to be compile-time constants), and we don't want people to omit the `const` now and get surprised later.

**Comment by leafpetersen** [comment]
Changed the name to reflect the outcome of the discussion.
